### DISCUSSION TITLE
DKI constraints from scratch

### DISCUSSION THREAD

**Discussion by samcoveney** [dipy/dipy#3419] [1. Ideas & Feedback]
Hi DiPy,

Related: #3222

So, Tom Dela Haije's work "Enforcing necessary non-negativity constraints for common diffusion MRI models using sum of squares programming" is behind the constraints for DKI fitting.

Tom provided, in npz files, the constraint matrices for DiPy. I have been in contact with Tom (who's been very helpful) and working from scratch myself, to reproduce these constraint matrices, which I appear to have successfully done - in fact, mine seem to be slightly more general than the ones Tom provided (my matrix L(\alpha) in eq (22) has 18 \alpha, whereas Tom's has only 9, suggesting that the linear space is not fully parametrized in the current DiPy constraint matrices - I am going to chase Tom up about this).

I have reproduced the constrained matrices by used "Sage Jupyter Notebook". It would be nice to somehow have the ability to produce constraint matrices in DiPy, but given how I have done it, I am not sure of the best way to proceed.

Thoughts on this?

**Comment by arokem** [comment]
Can you say a bit more about the code that you have written? I am not sure what "Sage" means in this context. Is it using this the ["sagemath"](https://www.sagemath.org/) library? Are you able to share the work that you have done so far so we can take a look and assess?

**Reply by samcoveney** [reply]
Yes the sagemath library seems to be what I have used.

So the DKI constraints in  "Enforcing necessary non-negativity constraints for common diffusion MRI models using sum of squares programming" take the form of a matrix containing the DKI model parameters \omega and slack variables \alpha. These matrices are loaded up here:

https://github.com/dipy/dipy/blob/927246108f6aa67283003a46b63f2deeced9becb/dipy/data/__init__.py#L410

I have recreated a set of constraint matrices that give the same results as the ones in DiPy. For technical reasons related to the ordering of basis monomials, and other such things relating to parameterization of the constraint matrix, the sets of matrices are not identical to the ones currently in DiPy, but appear to be doing exactly the same thing

What I have done, is to use sagemath to make a Jupyter notebook, which can then be run to produce these matrices. In my own branch of DiPy, I have swapped out the original constrain matrices for mine, and everything appears to work just fine.

It may be easier that I just make this notebook available somewhere, outside of DiPy, with no particular aim to having it be integrated into DiPy explicitly. What do you think?

**Comment by arokem** [comment]
Yeah - if the dependencies are very heavy that might be a reasonable approach. But I'd have to see the code to determine whether that's the case. For example, I believe that sympy is a reasonable dependency to take on, which may give you similar functionality. It depends a bit on whether this is something we think we might want to develop further.

**Comment by samcoveney** [comment]
https://github.com/samcoveney/convexity_matrix/blob/main/DKI_constraints.ipynb

This was a lot of work to figure out and do. Besides the need to tidy up, it's not a bad educational resource, in the sense that it's extremely explicit in how the matrices are actually figured out.

**Reply by samcoveney** [reply]
I am writing a paper that will include this, perhaps I can just reference the paper in DiPy for the explicit formulation?

**Reply by arokem** [reply]
That would be great. I think that including the code as a supplement to the paper would improve the paper, but that's a separate issue. Having read the code, I think that it would be hard to incorporate this in DIPY itself. That said, I agree with you that this could be a useful resource for teaching about this topic. I would suggest to add a bit more text that explains the calculations if that is the goal. Thanks!

**Reply by samcoveney** [reply]
No worries. An extra point that I need to verify, is that I think the matrices I have derived are slightly more general than the ones in DiPy currently, though I have only supplied the matrices for DKI and not DTI (we're not using the DTI ones anywhere at the moment - we could, which would guarantee PSD tensor with a least squares fit approach comparable to a DKI constrained fit approach, but frankly I don't see the point when non-linear fitting with Cholesky representation is easy for constrained DTI).

**Comment by samcoveney** [comment]
Happy to close this.
